waiter/src/waiter/metrics_sync.clj

```diff
@@ -206,31 +207,46 @@
   (with-catch
     router-metrics-state
     (let [my-router-id (:router-id router-metrics-state)
-          known-router-ids (disj (set (keys router-id->http-endpoint)) my-router-id)
-          prev-incoming-router-ids (set (keys router-id->incoming-ws))
-          prev-outgoing-router-ids (set (keys router-id->outgoing-ws))]
+          known-router-ids (disj (utils/keyset router-id->http-endpoint) my-router-id)
+          prev-incoming-router-ids (utils/keyset router-id->incoming-ws)
+          prev-outgoing-router-ids (utils/keyset router-id->outgoing-ws)]
       (if (or (not= known-router-ids prev-incoming-router-ids) (not= known-router-ids prev-outgoing-router-ids))
         (let [router-id->incoming-ws' (cleanup-router-requests :router-id->incoming-ws known-router-ids encrypt router-metrics-state)
               router-id->outgoing-ws' (cleanup-router-requests :router-id->outgoing-ws known-router-ids encrypt router-metrics-state)
-              new-outgoing-router-ids (->> (:router-id->outgoing-ws router-metrics-state) (keys) (set) (set/difference known-router-ids))]
-          (cid/cinfo "metrics-router-syncer" "new routers:" new-outgoing-router-ids)
-          (doseq [router-id new-outgoing-router-ids]
-            (let [ws-endpoint (-> (get router-id->http-endpoint router-id)
-                                  (str/replace "http://" "ws://")
-                                  (str "waiter-router-metrics"))
-                  request-id (str "inter-router-metrics-" (utils/unique-identifier))]
-              (cid/cinfo request-id "connecting to" router-id "at" ws-endpoint)
-              (ws/connect! websocket-client ws-endpoint
-                           (fn register-outgoing-request [{:keys [out] :as ws-request}]
-                             (let [ws-request (assoc ws-request :request-id request-id)]
-                               (async/>!! out (encrypt {:dest-router-id router-id
-                                                        :request-id request-id
-                                                        :source-router-id my-router-id
-                                                        :tag :initiate}))
-                               (send router-metrics-agent register-router-ws :router-id->outgoing-ws router-id
-                                     ws-request encrypt router-metrics-agent)))
-                           connect-options)))
-          (-> (preserve-metrics-from-routers router-metrics-state (set/union #{my-router-id} (set (keys router-id->http-endpoint))))
+              new-outgoing-router-ids (->> router-metrics-state
+                                           :router-id->outgoing-ws
+                                           utils/keyset
+                                           (set/difference known-router-ids))]
+          (when (not-empty new-outgoing-router-ids)
```

Comment:
Idiomatic is `(when (seq new-outgoing-router-ids) ...)`, and avoids requiring a collection.